pkg/kv-cache/kv-block/in-memory.go

```diff
@@ -0,0 +1,200 @@
+type PodCache struct {
+	// cache is an LRU cache that maps PodEntry to their last access time.
+	// thread-safe.
+	cache *lru.Cache[PodEntry, time.Time]
+}
+
+// Lookup receives a list of keys and a set of pod identifiers,
+// and retrieves the filtered pods associated with those keys.
+// The filtering is done based on the pod identifiers provided.
+// If the podIdentifierSet is empty, all pods are returned.
+//
+// It returns:
+// 1. A slice of the hit keys.
+// 2. A map where the keys are those in (1) and the values are pod-identifiers.
+// 3. An error if any occurred during the operation.
+func (m *InMemoryIndex) Lookup(ctx context.Context, keys []Key,
+	podIdentifierSet sets.Set[string],
+) ([]Key, map[Key][]string, error) {
+	if len(keys) == 0 {
+		return nil, nil, fmt.Errorf("no keys provided for lookup")
+	}
+
+	traceLogger := klog.FromContext(ctx).V(5).WithName("kvblock.InMemoryIndex.Lookup")
+
+	podsPerKey := make(map[Key][]string)
+	highestHitIdx := 0
+
+	for idx, key := range keys {
+		if pods, found := m.data.Get(key); found { //nolint:nestif // TODO: can this be optimized?
+			if pods == nil || pods.cache.Len() == 0 {
+				traceLogger.Info("no pods found for key, cutting search", "key", key.String())
+				return keys[:idx], podsPerKey, nil // early stop since prefix-chain breaks here
+			}
+
+			highestHitIdx = idx
+
+			if podIdentifierSet.Len() == 0 {
+				// If no pod identifiers are provided, return all pods
+				podsPerKey[key] = append(podsPerKey[key],
+					utils.SliceMap(pods.cache.Keys(), func(pod PodEntry) string {
+						return pod.PodIdentifier
+					})...)
+			} else {
+				// Filter pods based on the provided pod identifiers
+				for _, pod := range pods.cache.Keys() {
+					if podIdentifierSet.Has(pod.PodIdentifier) {
+						podsPerKey[key] = append(podsPerKey[key], pod.PodIdentifier)
+					}
+				}
+			}
+		} else {
+			traceLogger.Info("key not found in index", "key", key.String())
+		}
+	}
+
+	traceLogger.Info("lookup completed", "highest-hit-index", highestHitIdx,
+		"pods-per-key", podsPerKey)
+	return keys[:highestHitIdx+1], podsPerKey, nil
+}
+
+// Add adds a set of keys and their associated pod entries to the index backend.
+func (m *InMemoryIndex) Add(ctx context.Context, keys []Key, entries []PodEntry) error {
+	if len(keys) == 0 || len(entries) == 0 {
+		return fmt.Errorf("no keys or entries provided for adding to index")
+	}
+
+	traceLogger := klog.FromContext(ctx).V(5).WithName("kvblock.InMemoryIndex.Add")
+
+	for _, key := range keys {
+		podCache, found := m.data.Get(key) // bumps LRU timestamp if found
+		if !found {
+			cache, err := lru.New[PodEntry, time.Time](m.podCacheSize)
+			if err != nil {
+				return fmt.Errorf("failed to create pod cache for key %s: %w", key.String(), err)
+			}
+
+			podCache = &PodCache{
+				cache: cache,
+			}
+		}
+
+		for _, entry := range entries {
+			podCache.cache.Add(entry, time.Now()) // TODO: can this be batched to avoid multiple locks?
```

Comment:
you could change the interface to send an array of entries at once?

Reply:
`::Add` here is from hashicorp's `lru` package which does not expose batched addition. On every call an internal mutex is held, which could be a significant downside to using this implementation - investigation will follow up.